akka-http-core/src/main/scala/akka/http/impl/engine/parsing/ParserOutput.scala

```diff
@@ -64,7 +65,11 @@ private[http] object ParserOutput {
   sealed abstract class EntityCreator[-A <: ParserOutput, +B >: HttpEntity.Strict <: HttpEntity] extends (Source[A, Unit] ⇒ B)
 
   final case class StrictEntityCreator(entity: HttpEntity.Strict) extends EntityCreator[ParserOutput, HttpEntity.Strict] {
-    def apply(parts: Source[ParserOutput, Unit]) = entity
+    def apply(parts: Source[ParserOutput, Unit]) = {
+      // We might need to drain stray empty tail streams which will be read by no one.
+      parts.runWith(Sink.ignore)(GraphInterpreter.currentInterpreter.subFusingMaterializer)
```

Comment:
Thanks for the analysis, @drewhk, although it is slightly outdated in one regard: creating a substream in a GraphStage is now—while not completely trivial—quite simple actually (pending a PR from my side; example is the new impl of PrefixAndTail):
~~~scala
    private def subHandler = new OutHandler {
      override def onPull(): Unit = {
        cancelTimer(SubscriptionTimer)
        pull(in)
        tailSource.setHandler(new OutHandler {
          override def onPull(): Unit = pull(in)
        })
      }
    }

    private def openSubstream(): Source[T, Unit] = {
      val timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
      tailSource = new SubSourceOutlet[T]("TailSource")
      tailSource.setHandler(subHandler)
      scheduleOnce(SubscriptionTimer, timeout)
      builder = null
      Source.fromGraph(tailSource.source)
    }
~~~
The only sightly complex aspect is to not forget about the substream subscription timeout.

So, for me the conclusion is pretty clear, we’ll need to move this into the parser after 2.0.